cpu/cc2538: cleanup init, add cc2538_rf_obs_sig module

fjmolinas opened the pull request:
### Contribution description

This PR cleans up the init function of the `cc2538_rf` radio module, only non default parameters are set during the init procedure
and observable signals are made an optional modules. There is one important change where the radio does not automatically switch from `TX` to `RX`.

### Testing procedure

- `examples/gnrc_networking` should still work
- `tests/iee802154_hal` should still work

benpicco commented:
Looks good, but Travis found some issues.

fjmolinas commented:
I think I fixed what travis pointed out... lets see

benpicco commented:
I added `USEMODULE += cc2538_rf cc2538_rf_obs_sig` to the `Makefile` of `examples/gnrc_networking`.

Now I get a lot

```
2020-10-21 18:25:02,312 # cc2538_rf: Frame has not finished being received
2020-10-21 18:25:03,462 # cc2538_rf: Frame has not finished being received
2020-10-21 18:25:09,494 # cc2538_rf: Frame has not finished being received
```

on the `openmote-b`.

fjmolinas commented:
> I added `USEMODULE += cc2538_rf cc2538_rf_obs_sig` to the `Makefile` of `examples/gnrc_networking`.
> 
> Now I get a lot
> 
> ```
> 2020-10-21 18:25:02,312 # cc2538_rf: Frame has not finished being received
> 2020-10-21 18:25:03,462 # cc2538_rf: Frame has not finished being received
> 2020-10-21 18:25:09,494 # cc2538_rf: Frame has not finished being received
> ```
> 
> on the `openmote-b`.

Thanks for testing! I'll take a look

fjmolinas commented on cpu/cc2538/radio/cc2538_rf.c:
```diff
@@ -91,70 +100,75 @@ bool cc2538_channel_clear(void)
+    RFCORE_XREG_FIFOPCTRL = CC2538_RF_MAX_DATA_LEN;
+
+    /* Set default IRQ */
+    RFCORE_XREG_RFIRQM0 = RXPKTDONE | SFD | FIFOP;
+    if (IS_USED(MODULE_IEEE802154_RADIO_HAL)) {
+        RFCORE_XREG_RFIRQM1 = TXDONE | CSP_STOP;
     }
+    else {
+        RFCORE_XREG_RFIRQM1 = TXDONE;
+    }
```
I found the issue, I snuck in enabling two extra ISR from a follow up PR.

fjmolinas commented:
@benpicco it should be fine now, I had unadervtingly enabled `SFD` ISR.

benpicco commented on boards/openmote-b/include/board.h:
```diff
@@ -77,6 +77,15 @@
+ * @name    RF CORE observable signals settings
+ * @{
+ */
+#define CONFIG_CC2538_RF_OBS_SIG_0_PCX  0   /* PC4 */
+#define CONFIG_CC2538_RF_OBS_SIG_1_PCX  1   /* PC6 */
+#define CONFIG_CC2538_RF_OBS_SIG_2_PCX  2   /* PC7 */
```
The blinkenlights are not working on `openmote-b` but I can now use `at86rf215_subghz` together with `cc2538_rf`

fjmolinas replied:
I didn't adapt the pins... hahha

fjmolinas replied:
yeah I fixed that and something else that was mistaken in the init, but for some reason I currently only see the red led... dbugging

fjmolinas replied:
There was a missing include as well.

fjmolinas replied:
I know get flashy leds.

benpicco commented on cpu/cc2538/include/cc2538_rf.h:
```diff
@@ -212,7 +214,7 @@ enum {
     pa_pd            = 0x20,
     lna_pd           = 0x2a,
-    disabled         = 0x2b,
+    disabled,
```
All the other enum members are numbered explicitly, why remove it here?

fjmolinas replied:
The other all actually have that value, I though it might be clearer that that one was to be used for disabling, but I don't care either way.

benpicco replied:
you could have `disabled = 0xff`

benpicco commented on boards/openmote-b/include/board.h:
```diff
@@ -77,6 +77,15 @@
+ * @name    RF CORE observable signals settings
+ * @{
+ */
+#define CONFIG_CC2538_RF_OBS_SIG_0_PCX  4   /* PC4 */
```
```suggestion
#define CONFIG_CC2538_RF_OBS_SIG_0_PCX  5   /* PC5 */
```

PC4 is already LED0 which is the default LED used by applications

fjmolinas replied:
Will do.

benpicco reviewed (commented):
Now everything works.
The LEDs are a bit confusing - what are they supposed to indicate?

The Red `CONFIG_CC2538_RF_OBS_SIG_0_PCX` and Green `CONFIG_CC2538_RF_OBS_SIG_2_PCX` LED are always on.
If there is radio traffic, I can see `CONFIG_CC2538_RF_OBS_SIG_1_PCX` flashing.

If I put the radio to sleep, all LEDs will be on.

fjmolinas commented:
>  The LEDs are a bit confusing - what are they supposed to indicate?

I'm currently only keeping the current configuration. They are wired to `rx_active` `tx_active`, so you can see whether the radio is active/inactive I guess, when transmiting the one connected to `TX_ACTIVE` clearly blink. The other option is having them as `constant_value_1` (which is off) by default.

benpicco commented on cpu/cc2538/radio/cc2538_rf.c:
```diff
@@ -38,40 +38,50 @@
+static void _cc2538_observable_signals(void)
+{
+    if (IS_USED(MODULE_CC2538_RF_OBS_SIG)) {
+        if (CONFIG_CC2538_RF_OBS_0 != disabled) {
+            RFCORE_XREG_RFC_OBS_CTRL0 = CONFIG_CC2538_RF_OBS_0;
+            *(&CCTEST_OBSSEL0 + CONFIG_CC2538_RF_OBS_SIG_0_PCX) = \
+                CCTEST_OBSSELX_EN | rfc_obs_sig0;
+        }
+        if (CONFIG_CC2538_RF_OBS_1 != disabled) {
+            RFCORE_XREG_RFC_OBS_CTRL1 = CONFIG_CC2538_RF_OBS_1;
+            *(&CCTEST_OBSSEL0 + CONFIG_CC2538_RF_OBS_SIG_1_PCX) = \
+                CCTEST_OBSSELX_EN | rfc_obs_sig1;
+        }
+        if (CONFIG_CC2538_RF_OBS_2 != disabled) {
+            RFCORE_XREG_RFC_OBS_CTRL2 = CONFIG_CC2538_RF_OBS_2;
+            *(&CCTEST_OBSSEL0 + CONFIG_CC2538_RF_OBS_SIG_2_PCX) = \
+                CCTEST_OBSSELX_EN | rfc_obs_sig2;
+        }
```
Can you add a comment here what those signals indicate?

fjmolinas replied:
This logic just selects which observable signal goes to what pin, what each signal inficates depends on what is assigned to `CONFIG_CC2538_RF_OBS_%`. I'll add a comment.

fjmolinas commented:
> I'm currently only keeping the current configuration. They are wired to rx_active tx_active, so you can see whether the radio is active/inactive I guess, when transmiting the one connected to TX_ACTIVE clearly blink. The other option is having them as constant_value_1 (which is off) by default.

Or I could map other signals by defult if you think its more accurate?

fjmolinas commented:
@benpicco so what do you suggest to have as the default observed signals?

benpicco commented:
I don't know what they all mean :sweat_smile: 
But `rssi_valid` instead of `ffctrl_fifo` (what is that even?) gives a nice indication of an RXed packet

fjmolinas commented:
All options:

```
00 0000: 0 - Constant value
00 0001: 1 - Constant value
00 1000: rfc_sniff_data - Data from packet sniffer. Sample data on
rising edges of sniff_clk.
00 1001: rfc_sniff_clk - 250kHz clock for packet sniffer data.
00 1100: rssi_valid - Pin is high when the RSSI value has been
updated at least once since RX was started. Cleared when leaving
RX.
00 1101: demod_cca - Clear channel assessment. See FSMSTAT1
register for details on how to configure the behavior of this signal.
00 1110: sampled_cca - A sampled version of the CCA bit from
demodulator. The value is updated whenever a SSAMPLECCA or
STXONCCA strobe is issued.
00 1111: sfd_sync - Pin is high when a SFD has been received or
transmitted. Cleared when leaving RX/TX respectively. Not to be
confused with the SFD exception.
01 0000: tx_active - Indicates that FFCTRL is in one of the TX
states. Active-high.
Note: This signal might have glitches, because it has no output flip-
flop and is based
on the current state register of the FFCTRL FSM.
01 0001: rx_active - Indicates that FFCTRL is in one of the RX
states. Active-high.
Note: This signal might have glitches, because it has no output flip-
flop and is based
on the current state register of the FFCTRL FSM.
01 0010: ffctrl_fifo - Pin is high when one or more bytes are in the
RXFIFO. Low during RXFIFO overflow.
01 0011: ffctrl_fifop - Pin is high when the number of bytes in the
RXFIFO exceeds the programmable threshold or at least one
complete frame is in the RXFIFO. Also high
during RXFIFO overflow. Not to be confused with the FIFOP
exception.
01 0100: packet_done - A complete frame has been received. I.e.,
the number of bytes set by the frame-length field has been received.
01 0110: rfc_xor_rand_i_q - XOR between I and Q random outputs.
Updated at 8 MHz.
01 0111: rfc_rand_q - Random data output from the Q channel of
the receiver. Updated at 8 MHz.
01 1000: rfc_rand_i - Random data output from the I channel of the
receiver. Updated at 8 MHz
01 1001: lock_status - 1 when PLL is in lock, otherwise 0
10 1000: pa_pd - Power amplifier power-down signal
10 1010: lna_pd - LNA power-down signal
```

benpicco commented:
Would be nice if you could add those comments to the [`RFCORE_XREG_RFC_OBS_CTRLx` enum](https://github.com/RIOT-OS/RIOT/blob/d2526befa227b6464a23432e49b86322c6ca0797/cpu/cc2538/include/cc2538_rf.h#L197) :innocent:

benpicco commented on cpu/cc2538/include/cc2538_rf.h:
```diff
@@ -185,27 +195,83 @@ enum {
+#define CONFIG_CC2538_RF_OBS_1      rx_active
+#endif
+#ifndef CONFIG_CC2538_RF_OBS_2
+#define CONFIG_CC2538_RF_OBS_2      packet_done
```
`packet_done` is also always on for me.
I guess it never gets invalidated.
How about `rssi_valid`?

benpicco reviewed (approved):
Please squash.

fjmolinas commented:
@benpicco I had to rebase to fix a conflict

benpicco commented:
Looks like something broke during the rebase

fjmolinas commented:
> Looks like something broke during the rebase

An ifdef snuck in not sure how.

fjmolinas commented:
All green!

aabadie commented:
Go!

fjmolinas commented:
Thanks for the review @benpicco!
